src/software/util/design_patterns/generic_factory.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+// TODO should be removed later
+#include <functional>
+//#include "software/backend/backend.h" // Should make generic.h
+
+
+// A quality of life typedef to make things shorter and more readable
+template <class GenericType>
+        using GenericRegistry = std::unordered_map<std::string, std::function<std::unique_ptr<GenericType>()>>;
```

Comment:
Could we change this to be `template <class IndexType, class GenericType> using GenericRegistry = std::unordered_map<IndexType, std::function<std::unique_ptr<GenericType>()>>;`? 

That way we can use things like enums to do lookups in the future.

Reply:
Changed to use that.